### PR TITLE
Adds initial integration with Zendesk

### PR DESCRIPTION
**Pull request by mike-lvov**

### CODE REVIEW

**Comment by thecadams** [comment]
Hi @mike-lvov @ thanks for the contribution. Just wondering, what's the difference between this approach and including zendesk directly in `index.html`? Are there advantages to doing it in React like this?

File: src/index.tsx [hunk 1]
```diff
@@ -54,6 +54,7 @@ import { User } from "types/User";
 
 import { useSelector } from "hooks/useSelector";
 import { authSelector } from "utils/selectors";
+import { initializeZenDesk } from "utils/zenDesk";
```
Comment (thecadams) [inline src/index.tsx:57]: nit: `Zendesk` - no capital D. (https://www.sec.gov/Archives/edgar/data/1463172/000119312514137878/d665368ds1.htm)

File: src/utils/zenDesk.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+const ZENDESK_SCRIPT_SRC =
+  "https://static.zdassets.com/ekr/snippet.js?key=57ad1fbc-d174-4561-8102-b3c8d98436d5";
+
+const ZENDESK_SCRIPT_ID = "ze-snippet";
+
+export function initializeZenDesk() {
+  loadScript(ZENDESK_SCRIPT_SRC, () => {}, {
+    id: ZENDESK_SCRIPT_ID,
+  });
+}
```
Comment (thecadams) [inline src/utils/zenDesk.ts:12]: nite: per above nit, rename this file to `zendesk.ts`.

File: src/utils/loadScript.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { HTMLAttributes } from "react";
+
+export function loadScript(
+  src: string,
+  onLoad: () => void,
```
Comment (thecadams) [inline src/utils/loadScript.ts:5]: nit: considering it's going into a property called `onload`, is this better off being called `onload` instead of `onLoad`? Open to opinions here.

**Review by thecadams — COMMENTED** [review]
Looking good, pending my higher level comment on the approach.

**Comment by thecadams** [comment]
From Mike:
This way lets us pass in functions to notify when the script is loaded so we have more control for developers. So this is a superior way to do this one.

File: src/utils/loadScript.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+type ScriptProps = "src" | "id";
+type LoadScriptProps = Record<ScriptProps, string>;
+
+export function loadScript(options: LoadScriptProps) {
```
Comment (denisdimitrov) [inline src/utils/loadScript.ts:4]: Is there a reason for the ES5 syntax? On the next line you are using `const` I suppose this can become an arrow function and keep it consistent.

**Review by denisdimitrov — CHANGES_REQUESTED** [review]

File: src/utils/loadScript.ts (hunk 4 above)
Comment (denisdimitrov) [reply src/utils/loadScript.ts:4]: Same question/logic applies for the function `initializeZendesk`

Comment (mike-lvov) [reply src/utils/loadScript.ts]: Good addition, changed that

**Comment by mike-lvov** [comment]
> Hi @mike-lvov & thanks for the contribution. Just wondering, what's the difference between this approach and including zendesk directly in `index.html`? Are there advantages to doing it in React like this?

I think this is a really good question. As far as I understand, there will be more added to this integration and I think it is useful to have all of these things in one place/file

**Review by denisdimitrov — APPROVED** [review]
LGTM!

**Review by thecadams — APPROVED** [review]
LGTM!

File: src/utils/zendesk.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { loadScript } from "./loadScript";
+
+const ZENDESK_SCRIPT_SRC =
+  "https://static.zdassets.com/ekr/snippet.js?key=57ad1fbc-d174-4561-8102-b3c8d98436d5";
```
Comment (IvanBuljovcic) [inline src/utils/zendesk.ts:4]: nit: Should this be in here or `settings.ts` ?

File: src/utils/zendesk.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+const ZENDESK_SCRIPT_ID = "ze-snippet";
```
Comment (IvanBuljovcic) [inline src/utils/zendesk.ts:6]: same as above

File: src/utils/zendesk.ts (hunk 5 above)
Comment (0xdevalias) [reply src/utils/zendesk.ts:4]: Personally, I would have implemented it as an environment variable that is then mapped through, rather than as a hardcoded static. That method aligns more to how we handle existing things like this, and makes the code more flexible when others are deploying their own versions of it.

This would be mapped in https://github.com/sparkletown/sparkle/blob/staging/src/secrets.js (which isn't necessarily the best filename, since most of these values aren't actually secret.. particularly when mapped to the frontend.. might be worth renaming it `env` or similar? Realistically.. this file is actually what most of `settings.ts` would end up as one day I think.. so maybe it should be merged there or similar?)

Comment (0xdevalias) [reply src/utils/zendesk.ts:4]: Backlogged these improvements in https://github.com/sparkletown/internal-sparkle-issues/issues/100

File: src/utils/zendesk.ts (hunk 6 above)
Comment (0xdevalias) [reply src/utils/zendesk.ts:6]: See https://github.com/sparkletown/internal-sparkle-issues/issues/100

**Comment by 0xdevalias** [comment]
> @thecadams
> From @mike-lvov:
> This way lets us pass in functions to notify when the script is loaded so we have more control for developers. So this is a superior way to do this one.

> @mike-lvov
> I think this is a really good question. As far as I understand, there will be more added to this integration and I think it is useful to have all of these things in one place/file

It's been a while since I had my 🧠  in this feature/the documentation/etc, but from memory I the conclusion I came to when planning it out was that loading the script in the HTML, and then adding the relevant JS/etc parts to `index.ts`/it's own file that is imported/etc would be the 'best' solution here (as documented on https://github.com/sparkletown/internal-sparkle-issues/issues/65#issuecomment-741539135 + the related slack link there):

> tl;dr: `@tech-leads` i’m going to integrate this with the raw zendesk snippet + maybe create our own helpers to wrap around their ‘raw API’ in a typescript friendly way.

If the current method works and doesn't add any inefficiencies/potential issues to our codebase, then happy to leave it as is for now. But just wanted to capture this here for potential future cleanup/knowledge/etc.
